src/bandersnatch/mirror.py

```diff
@@ -193,6 +109,139 @@ def _filter_packages(self) -> None:
                 else:
                     del self.packages_to_sync[package_name]
 
+    async def determine_packages_to_sync(self) -> None:
+        """
+        Update the self.packages_to_sync to contain packages that need to be
+        synced.
+        """
+        raise NotImplementedError()
+
+    async def package_syncer(self, idx: int) -> None:
+        logger.debug(f"Package syncer {idx} started for duty")
+        while True:
+            try:
+                package = self.package_queue.get_nowait()
+                await package.update_metadata(self.master, attempts=3)
+                await self.process_package(package)
+            except asyncio.QueueEmpty:
+                logger.debug(f"Package syncer {idx} emptied queue")
+                break
+            except PackageNotFound:
+                continue
+            except Exception as e:
+                self.on_error(e, package=package)
+
+    async def process_package(self, package: Package) -> None:
+        raise NotImplementedError()
+
+    async def sync_packages(self) -> None:
+        try:
+            self.package_queue: asyncio.Queue = asyncio.Queue()
+            # Sorting the packages alphabetically makes it more predictable:
+            # easier to debug and easier to follow in the logs.
+            for name in sorted(self.packages_to_sync):
+                serial = int(self.packages_to_sync[name])
+                await self.package_queue.put(Package(name, serial=serial))
+
+            sync_coros: List[Awaitable] = [
+                self.package_syncer(idx) for idx in range(self.workers)
+            ]
+            try:
+                await asyncio.gather(*sync_coros)
+            except KeyboardInterrupt as e:
+                self.on_error(e)
+        except (ValueError, TypeError) as e:
+            # This is for when self.packages_to_sync isn't of type Dict[str, int]
```

Comment:
Does this happen as part of the normal control flow?  Under what circumstances?  If this happens semi-regularly, then making a distinguished exception type would be preferable I think.

But I'm missing some context here

Reply:
Actually this error check is for some poorly written tests.  No where in the normal control flow should this error occur. It happens when one manually sets the self.packages_to_sync to any value that doesn't follow the type Dict[str, Union[int, str]], which some of the tests do purposely set to incorrect values.

Reply:
Oof.  Let's definitely call that out as a TODO then.

As an aside, that doesn't really make sense, the entire point of adding type annotations is to catch bugs like that so that unit tests for that sort of thing are no longer necessary.  So hopefully it can be ditched.